lib/github.dart

```diff
@@ -0,0 +1,343 @@
+/// exists, or `GITHUB_TOKEN` otherwise.
+String get pkgGithubPassword {
+  _pkgGithubPassword ??= Platform.environment["GITHUB_PASSWORD"] ??
+      Platform.environment["GITHUB_TOKEN"];
+  if (_pkgGithubPassword != null) return _pkgGithubPassword;
+
+  fail("pkgGithubPassword must be set to deploy to GitHub.");
+  return null; // google/grinder.dart#339
+}
+
+set pkgGithubPassword(String value) => _pkgGithubPassword = value;
+String _pkgGithubPassword;
+
+/// Returns the HTTP basic authentication Authorization header from the
+/// environment.
+String get _authorization =>
+    "Basic ${base64.encode(utf8.encode("$pkgGithubUser:$pkgGithubPassword"))}";
+
+/// The Markdown-formatted release notes to use for the GitHub release.
+///
+/// By default, this looks for a `CHANGELOG.md` file at the root of the
+/// repository and uses the portion between the first and second level-2 `##`
+/// headers. It throws a [FormatException] if the CHANGELOG doesn't begin with
+/// `##` followed by [pkgVersion].
+String get pkgGithubReleaseNotes {
+  if (_pkgGithubReleaseNotes != null) return _pkgGithubReleaseNotes;
+  if (!File("CHANGELOG.md").existsSync()) return null;
+
+  _pkgGithubReleaseNotes = _lastChangelogSection() +
+      "\n\n"
+          "See the [full changelog](https://github.com/$pkgGithubRepo/blob/"
+          "master/CHANGELOG.md#${pkgVersion.toString().replaceAll(".", "")}) "
+          "for changes in earlier releases.";
+  return _pkgGithubReleaseNotes;
+}
+
+set pkgGithubReleaseNotes(String value) => _pkgGithubReleaseNotes = value;
+String _pkgGithubReleaseNotes;
+
+/// Creates a GitHub release for [pkgVersion] of this package.
+///
+/// This creates the release on the [pkgGithubRepo] repository, using
+/// [pkgHumanName] as the release name and [pkgGithubReleaseNotes] as the
+/// release notes.
+@Task('Create a GitHub release, without executables.')
+Future<void> pkgGithubRelease({http.Client client}) async {
+  var response = await withClient(client, (client) {
+    return client.post(
+        url("https://api.github.com/repos/$pkgGithubRepo/releases"),
+        headers: {
+          "content-type": "application/json",
+          "authorization": _authorization
+        },
+        body: jsonEncode({
+          "tag_name": pkgVersion.toString(),
+          "name": "$pkgHumanName $pkgVersion",
+          "prerelease": pkgVersion.isPreRelease,
+          if (pkgGithubReleaseNotes != null) "body": pkgGithubReleaseNotes
+        }));
+  });
+
+  if (response.statusCode != 201) {
+    fail("${response.statusCode} error creating release:\n${response.body}");
+  } else {
+    log("Released $pkgHumanName $pkgVersion to GitHub.");
+  }
+}
+
+/// A regular expression that matches a Markdown code block.
+final _codeBlock = RegExp(" *```");
+
+/// Returns the most recent section in the CHANGELOG, reformatted to remove line
+/// breaks that will show up on GitHub.
+String _lastChangelogSection() {
```

Comment:
A lot of this method seems fragile, but you don't want to write a full markdown parser :/

E.g. I can see some projects using h1 (`#`); there might be less than 1 space or more  than one space  after `##`. Probably not too important until you get some bugs filed 😄

Reply:
Yeah, given that users have the escape hatch of manually setting `pkgGithubReleaseNotes`, I think it's fine to wait on making this super customizable.